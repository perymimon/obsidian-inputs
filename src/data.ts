// update cache with inline-field meta data
import type {TFile} from "obsidian";
import {Priority, CachedStructure, targetFile} from "./types";
import {getTFile, lastCreatedFiles, lastTouchFiles} from "./files";
import {setPrototype} from "./basics/objects";
import {getInlineFields} from "./data.inlineFields";

/**
 * @param file
 * @param priority 'yaml'|'field'
 */
const app = globalThis.app;

export function getFileData(file?: targetFile, extraVars = {}, priority: Priority = 'field') {
	const context: any = {}
	for (let i in lastTouchFiles) context[`page${i}`] = lastTouchFiles[i]
	for (let i in lastCreatedFiles) context[`new${i}`] = lastCreatedFiles[i]
	let tFile = getTFile(file)
	if (!tFile) return context
	const {frontmatter = {}, inlineFields = {}, dirty} = getFileStructure(tFile)
	context.dirty = dirty
	if (priority == 'field') return setPrototype(extraVars, inlineFields, frontmatter, context)
	if (priority == 'yaml') return setPrototype(extraVars, frontmatter, inlineFields, context)
	return setPrototype(extraVars, inlineFields, frontmatter, context)
}

export function getFileStructure(file?: targetFile): CachedStructure {
	let tFile = getTFile(file)
	if (!tFile) return {dirty: true}
	var cache = app.metadataCache.getFileCache(tFile)
	if (!cache) return {dirty: true}
	return cache as CachedStructure
}


export async function waitFileStructureReady(tFile: TFile) {
	var time = 10
	do {
		await sleep(time)
		if (time > 2000) throw `more then 2s and the file ${tFile.path} not ready`
		var data = getFileStructure(tFile)
		time *= 2
	} while (data.dirty)
	return data
}

export async function refreshFileStructure(targetFile?: targetFile) {
	let tFile = getTFile(targetFile)
	let cache = await waitFileStructureReady(tFile)
	if (!cache) throw `No cache found. for ${tFile.path}`;

	const content = await app.vault.cachedRead(tFile)
	const inlineFields = getInlineFields(content)

	const fieldsObject: object = inlineFields.reduce(
		(obj, line) => (obj[line.key] = line.value, obj), {}
	)
	cache.allInlineFields = inlineFields
	cache.inlineFields = fieldsObject
	cache.dirty = false
	return cache as CachedStructure
}


