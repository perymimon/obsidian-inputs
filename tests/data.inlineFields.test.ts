import {getInlineFields} from "../src/data.inlineFields";


describe('getInlineFields from bulk of content', () => {
	test('one field [variable::value] in a row', async () => {
		const array = getInlineFields(`
				# title
				some content [variable::value]
				extra line
			`)
		expect(array).toMatchObject([{
			"fullKey": "variable",
			"innerField": "variable::value",
			"isRound": true,
			"isSquare": false,
			"key": "variable",
			"keyOffset": [31, 39],
			"offset": [30, 47],
			"oldValue": "value",
			"outerField": "[variable::value]",
			"value": "value",
			"valueOffset": [41, 46],
		}])
	})
	const cases = [[
		'[variable::value] [variable::value]',
		`
				# title
				some content [variable::value] [variable::value]
				extra line
			`
	], [
		'with wikiLink as value [variable::[[value]]]',
		`
				# title
				some content [variable::[[value]]]
				extra line
			`
	], [
		'with wikiLink as value variable::[[value]] and more after text',
		`
			# title [::[[value]]]
			some content variable::[[value]] and more after text
			extra line
		`
	], [
		'with wikiLink as value variable::[[value]] and more after text (time::10)',
		`
			# title [::[[value]]]
			some content variable::[[value]] and more after text (time::10)
			extra line
		`
	]]

	for (const _case of cases) {
		test(_case[0], () => {
			const array = getInlineFields(_case[1])
			expect(array).toMatchSnapshot()
		})
	}


})
