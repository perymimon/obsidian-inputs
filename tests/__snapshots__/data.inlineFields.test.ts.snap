// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getInlineFields [variable::[[value]]] 1`] = `
[{
    "isRound": "true",
    "isSquare": "false",
    "outerField": "[variable::[[value]]]",
    "innerField": "variable::[[value]]",
    "key": "variable",
    "value": "[[value]]",
    "fullKey": "variable",
    "oldValue": "[[value]]",
    "offset": "[30,51]",
    "keyOffset": "[31,39]",
    "valueOffset": "[41,50]"
  }]
`;

exports[`getInlineFields [variable::value] [variable::value] 1`] = `
[{
    "isRound": "true",
    "isSquare": "false",
    "outerField": "[variable::value]",
    "innerField": "variable::value",
    "key": "variable",
    "value": "value",
    "fullKey": "variable",
    "oldValue": "value",
    "offset": "[30,47]",
    "keyOffset": "[31,39]",
    "valueOffset": "[41,46]"
  },{
    "isRound": "true",
    "isSquare": "false",
    "outerField": "[variable::value]",
    "innerField": "variable::value",
    "key": "variable",
    "value": "value",
    "fullKey": "variable",
    "oldValue": "value",
    "offset": "[48,65]",
    "keyOffset": "[49,57]",
    "valueOffset": "[59,64]"
  }]
`;

exports[`getInlineFields from bulk of content [variable::value] [variable::value] 1`] = `
[{
    "isRound": "true",
    "isSquare": "false",
    "outerField": "[variable::value]",
    "innerField": "variable::value",
    "key": "variable",
    "value": "value",
    "fullKey": "variable",
    "oldValue": "value",
    "offset": "[30,47]",
    "keyOffset": "[31,39]",
    "valueOffset": "[41,46]"
  },{
    "isRound": "true",
    "isSquare": "false",
    "outerField": "[variable::value]",
    "innerField": "variable::value",
    "key": "variable",
    "value": "value",
    "fullKey": "variable",
    "oldValue": "value",
    "offset": "[48,65]",
    "keyOffset": "[49,57]",
    "valueOffset": "[59,64]"
  }]
`;

exports[`getInlineFields from bulk of content with wikiLink as value [variable::[[value]]] 1`] = `
[{
    "isRound": "true",
    "isSquare": "false",
    "outerField": "[variable::[[value]]]",
    "innerField": "variable::[[value]]",
    "key": "variable",
    "value": "[[value]]",
    "fullKey": "variable",
    "oldValue": "[[value]]",
    "offset": "[30,51]",
    "keyOffset": "[31,39]",
    "valueOffset": "[41,50]"
  }]
`;

exports[`getInlineFields from bulk of content with wikiLink as value variable::[[value]] and more after text (time::10) 1`] = `
[{
    "isRound": "true",
    "isSquare": "false",
    "outerField": "[::[[value]]]",
    "innerField": "::[[value]]",
    "key": "",
    "value": "[[value]]",
    "fullKey": "",
    "oldValue": "[[value]]",
    "offset": "[12,25]",
    "keyOffset": "[13,13]",
    "valueOffset": "[15,24]"
  },{
    "isRound": "false",
    "isSquare": "false",
    "outerField": "variable::[[value]] and more after text (time::10)",
    "innerField": "variable::[[value]] and more after text (time::10",
    "key": "variable",
    "value": "[[value]] and more after text (time",
    "fullKey": "variable",
    "oldValue": "[[value]] and more after text (time",
    "offset": "[42,92]",
    "keyOffset": "[43,51]",
    "valueOffset": "[53,88]"
  },
  {
    "isRound": "false",
    "isSquare": "true",
    "outerField": "(time::10)",
    "innerField": "time::10",
    "key": "time",
    "value": "10",
    "fullKey": "time",
    "oldValue": "10",
    "offset": "[82,92]",
    "keyOffset": "[83,87]",
    "valueOffset": "[89,91]"
  }]
`;

exports[`getInlineFields from bulk of content with wikiLink as value variable::[[value]] and more after text 1`] = `
[{
    "isRound": "true",
    "isSquare": "false",
    "outerField": "[::[[value]]]",
    "innerField": "::[[value]]",
    "key": "",
    "value": "[[value]]",
    "fullKey": "",
    "oldValue": "[[value]]",
    "offset": "[12,25]",
    "keyOffset": "[13,13]",
    "valueOffset": "[15,24]"
  },{
    "isRound": "false",
    "isSquare": "false",
    "outerField": "variable::[[value]] and more after text",
    "innerField": "variable::[[value]] and more after text",
    "key": "variable",
    "value": "[[value]] and more after text",
    "fullKey": "variable",
    "oldValue": "[[value]] and more after text",
    "offset": "[42,81]",
    "keyOffset": "[42,50]",
    "valueOffset": "[52,81]"
  }]
`;

exports[`getInlineFields with wikiLink as value [variable::[[value]]] 1`] = `
[{
    "isRound": "true",
    "isSquare": "false",
    "outerField": "[variable::[[value]]]",
    "innerField": "variable::[[value]]",
    "key": "variable",
    "value": "[[value]]",
    "fullKey": "variable",
    "oldValue": "[[value]]",
    "offset": "[30,51]",
    "keyOffset": "[31,39]",
    "valueOffset": "[41,50]"
  }]
`;

exports[`getInlineFields with wikiLink as value variable::[[value]] and more after text 1`] = `
[{
    "isRound": "true",
    "isSquare": "false",
    "outerField": "[::[[value]]]",
    "innerField": "::[[value]]",
    "key": "",
    "value": "[[value]]",
    "fullKey": "",
    "oldValue": "[[value]]",
    "offset": "[12,25]",
    "keyOffset": "[13,13]",
    "valueOffset": "[15,24]"
  },{
    "isRound": "false",
    "isSquare": "false",
    "outerField": "variable::[[value]] and more after text",
    "innerField": "variable::[[value]] and more after text",
    "key": "variable",
    "value": "[[value]] and more after text",
    "fullKey": "variable",
    "oldValue": "[[value]] and more after text",
    "offset": "[42,81]",
    "keyOffset": "[42,50]",
    "valueOffset": "[52,81]"
  }]
`;
